### ISSUE TITLE
Brainstorm - New Runner Output Format

### ISSUE THREAD

**Issue by jhayley** [jhayley/bondage.js#18]
So, as it stands right now, the bondage runner has three types of outputs that are returned to the user, as defined in [`results.js`](https://github.com/jhayley/bondage.js/blob/master/src/results.js): `LineResult`, `OptionsResult`, and `CommandResult`. This isn't very extensible, but it works for the basic yarn functionality, and is basically everything the end user will care about in that specific case.

I am interested in making this more generic to allow for more complex formats to be returned easily from the AST, rather than worrying about what types they are on the runner side. 

In this situation I'm imaging three separate components, a Parser, a Runner, and a Renderer (the former two are currently encompassed by bondage). In the future, the Parser should be a separate module with a well-defined interface that can be plugged into bondage, while also allowing the option for other sorts of languages to be used as long as the provide the interface. The Renderer would simply plug into the runner and consume its output, handling all of the decisions of how to display and process it, as well as being responsible for reporting the status of any player branching decisions to the runner.

I think the ideal of this system would be one where
1) Any arbitrary amount of new commands/language features can be added without changing the runner (only the parser).
2) The runner only cares about options/conditionals/branching and variables, the rest can be shunted off to the renderer. It should handle all logic regarding these things, as well as keeping track of the player input as reported for the renderer (e.g. one theoretical runner feature would be the ability to reverse through the decisions the player has made back up the tree) 
3) Some sort of type definition can be provided by the parser to denote what sort of nodes exist, and what types they are, what properties they contain, etc. This can be checked by the renderer to ensure it knows what featureset it's dealing with. This could also be processed on the runner side to check the validity of the AST.

I think the easiest way to achieve points 1 and 2 is to have the runner pass AST data more-or-less directly to the renderer. Each AST object would contain a `type` string describing what it is, and then it would contain the parameters described in the type definition. The renderer would be able to handle the types and use them however it wants.  
There are a few exceptions to this: the parser would need to always define the features the runner needs (as mentioned in the beginning of point 2). Some sort of standard object type to define branches and variables that doesnt get sent to the renderer, but is used to determine the flow of the dialog. The renderer shouldn't even need to know what branches are being taken, it just displays any options given by the renderer, then tells the runner when they are selected (just as it doesn't need to know how to handle variables/scoping/calculations).

In my test visual novel renderer, I'm just running it directly based off of an extremely simple AST (with a small mocked up runner). Here is an excerpt of it:

```javascript
[
    {
      type: 'hide',
      name: 'me1',
      transition: 'fadeout',
    },
    {
      type: 'show',
      name: 'me2',
      tag: 'normal',
      loc: 'right',
      transition: 'fadein',
    },
    {
      type: 'say',
      text: 'lorem ipsum',
      name: 'hayley',
    },
    {
      type: 'hide',
      name: 'me2',
      transition: 'fadeout',
    },
    {
      type: 'option',
      options: [
        {
          text: 'This is option 1',
          name: 'option1',
        },
        {
          text: 'This is option 2',
          name: 'option2',
        },
      ],
      onChoose: (name) => {
        console.log(`Choice: ${name}`);
      },
    },
]
```

Note the onChoose function, this would be inserted by the runner (or just be a function on the runner itself rather than being put into the ast object) to be called by the renderer to signal when an option is selected.

Phew, that was a lot. Open to thoughts/suggestions. Hopefully doing something like this could allow for a variety of sorts of projects (twine-likes, vns, "full" games) to all use this runner.

**Comment by konistehrad** [comment]
Allow me to present: a snippet of the XML file that drives Starfighter: Eclipse!

```
   <ParallelAction auto="true" name="_1">
      <events>
        <AudioAction fadeDuration="1" target="drone" volume=".175"/>
        <ActorFade opacity="1" target="curtain"/>
      </events>
    </ParallelAction>
    <ActorFade auto="true" name="_2" duration="1" opacity="0" target="curtain"/>
    <!--Remember what I taught you?-->
    <TextAction name="_3" text="@t_sday1intro_0"/>
    <!--No matter what happens… You keep fighting.-->
    <TextAction name="_4" text="@t_sday1intro_1"/>
    <!--I’m waiting on you. You better catch up, little brother.-->
    <TextAction name="_5" text="@t_sday1intro_2"/>
    <ParallelAction auto="true" name="_6">
      <events>
        <AudioAction fadeDuration="2" target="drone" volume="0"/>
        <TextAction text=" "/>
        <ActorFade duration=".33" opacity="0" target="textArea"/>
      </events>
    </ParallelAction>
    <ParallelAction allowInterrupt="false" auto="true" name="_7">
      <events>
        <AudioAction delay=".75" fadeDuration="2" loop="true" play="true" sound="@a_sfx_airporttarmac_6061_41" target="drone" volume="0.5"/>
        <ActorFade delay="1" duration="1" opacity="1" target="cg"/>
      </events>
    </ParallelAction>
    <ChoiceAction name="_J" varName="day1_0">
      <choices>
        <!--Yes, I’m looking for my new Navigator, are you him?-->
        <choice value="0">@t_sday1intro_92</choice>
        <!--Mind your own business.-->
        <choice value="1">@t_sday1intro_93</choice>
      </choices>
    </ChoiceAction>
```

Some notes here: we need to be able to issue commands that run in parallel and are logically grouped. Some commands should be terminated early by snapping to the end on click, others should block input until they are complete. Some should automatically progress the flowchart, others should wait for input. Text should allow string-key references for internationalization. The names exist to provide a stable target for loading the games. Choices save values to variables.

Might be something helpful in here.

**Comment by jhayley** [comment]
Oooh thank you! I'm reassured by how similar your format looks to mine, and those are some really handy things I hadn't thought about

**Comment by jhayley** [comment]
Thinking about parallel actions more. The thing about them is that the *renderer* is the one that actually cares about doing the actions, so that would be its responsibility. But also, the renderer is the one responsible for requesting subsequent actions, so I think parallel actions could just be achieved by the parser having a flag that could be set just as "process next action immediately". Instead of defining a block of parallel actions, the renderer would know to just request another action right away from the runner instead of waiting,  which doesn't require any extra sorts of processing on the runner's part.

**Comment by jhayley** [comment]
Internationalization would need to be something on the parser side that can process it then just send text as an object like { 'en': 'asdf', 'eo' : 'blah' } instead of a string, then the renderer could be free to choose based on settings or whatever. Not sure how that would work specifically, maybe just a token in the language to describe the following text as a key reference rather than raw, which then gets inserted into the data before sending it to the runner.

**Comment by jhayley** [comment]
oh also @konistehrad, are you generating that xml from something?

**Comment by konistehrad** [comment]
> Thinking about parallel actions more

So this structure is the way that Fungus works presently; it can be difficult to logically group things for termination on click. I extended the Fungus `Say` command to act as a "sentinel" of sorts, with the rest of the system waiting on these sentinels on concurrently executing Commands to complete or waiting for input to send termination signals to those that are still executing. That said, this structure works well enough for Fungus so you might be right about that.

> Internationalization would need to be something on the parser side that can process it then just send text as an object like { 'en': 'asdf', 'eo' : 'blah' } instead of a string

Actually yes this would be lovely. Packing alternate languages into the "Say" commands directly would spare me the trouble of building XML's (htmlVN) or CSV's (Fungus) and would present the end user with immediate feedback as to their location in the script in their native language. Love it.

> are you generating that xml from something?

That's production game script from day1 of Starfighter: Eclipse. It was generated initially by parsing a google doc, and the actions were scripted in Eclipse (ironic, I know!). I never ended up completing the node-based editor. :<

**Comment by jhayley** [comment]
> it can be difficult to logically group things for termination on click

ahhh yes I can see how that would be. this would still be something that was handled on the renderer side, but the runner would need to handle sending blocks of commands to the renderer instead of one by one. Instead sending one by one it would be more "here's a list of things that need to be handled now"

**Comment by jhayley** [comment]
I build an extremelyyy basic text/option selecting renderer on top of the *current* bondage featureset as a proof of concept: https://github.com/jhayley/bondage-renderer-test (can see it running at http://hayley.zone/bondage-renderer-test/)

There is some weird bits in processing the output and sending it to the renderer in this case, but that's because of the issues with how bondage currently outputs each step.

The interface for the renderer is essentially just a `do` function that takes a type of event and some options, checks to see if that event is a valid action, then passes it on to a function that performs the action. I had a nice time implementing this, it was pretty straightforward all told so I think that's a good sign??

**Comment by ephread** [comment]
Very cool to see some ideas exchanged around!

I'm not sure I fully grasp of the extent of the discussion, as my experience with visual noveling is pretty limited, so feel free to correct me. I think we're mostly on the same page, but the whole discussion seemed to jumbled [_parser_ -> _runner_] and [_runner_ <-> _renderer_] together, which puzzled me a bit.

> **@jhayley**
> But also, the renderer is the one responsible for requesting subsequent actions, so I think parallel actions could just be achieved by the parser having a flag that could be set just as "process next action immediately".

I am confused about the above JSON excerpt (is it a simple AST, as in the AST outputted by the _parser_? or by the _runner_?), and that brings two (maybe out of scope) questions here:

- should the _renderer_ really be requesting actions, rather than simply reports their completion to the _runner_ (which would handle the flow) ?
- should the _runner_ and the _renderer_ really talk to each other through a complex JSON structure, rather than letting the _runner_ call the appropriate methods on the _renderer_ directly? Are there significant advantages?

I may not see the bigger picture here but I envisioned the renderer as something providing a simple, with a standardized javascript API, really like the work jhayley did in [`renderer.js`](https://github.com/jhayley/bondage-renderer-test/blob/master/src/renderer.js).

I, too, believe the _runner_ is where all the logical magic happens. It traverses the AST outputed by the _parser_ and when it reaches a text node, it tells the _renderer_ to show the text and wait for the _renderer_ to report completion. When it reaches an option link, it tells the _renderer_ to ask the user for feedback and report the result.

As for the parallel actions, if I understand the concept properly:

when the _runner_ stumble upon a parallel action node (PA), it resolves the child nodes of this PA and tell the _renderer_ to perform the contained actions at once and wait for the _renderer_ to report completion for all actions, before carrying on.

Different renderers would support different actions. For instance, the command line renderer would not support rendering sound, thus, the action would complete immediately as if it was performed. Or, maybe, as mentioned in the thread (I think), each renderer could provide a set of actions it supports and the runner would simply ignore certain nodes of the AST based on this knowledge.

**Comment by jhayley** [comment]
> should the renderer really be requesting actions, rather than simply reports their completion to the runner (which would handle the flow) ?

I think in this case we're referring to the same thing just describing it differently. I meant "request an action" in the sense of going to the runner and saying "I'm done give me another one". Thinking of it as reporting completion for the specified action makes more sense, and also lends more to the idea of things being reported back to the runner on completion (mostly of condition choices, but I'm sure there are other cases for sending data back)

> should the runner and the renderer really talk to each other through a complex JSON structure, rather than letting the runner call the appropriate methods on the renderer directly? Are there significant advantages?

This makes sense. The way I've been writing my test renderers was just consuming raw json because that's basically what bondage does right now so that's how I'm describing it, but you're totally right. Though thinking about it, there may be a key distinction here: this assumes that whenever an action completes, the renderer wants the next one immediately. If a renderer could also be a larger game, maybe after an option they wouldn't need to know the following action until later? This would put the onus on them to queue the next action and just not report its completion for a while, or to delay the previous action's completion until they need the next step. I think your way definitely works _better_ in most cases though, especially for parallel actions.

>when the runner stumble upon a parallel action node (PA), it resolves the child nodes of this PA and tell the renderer to perform the contained actions at once and wait for the renderer to report completion for all actions, before carrying on.

i like this, it's a pretty graceful way to put it actually. Just tying it to the completion report mechanism means the renderer doesn't need to change its behavior and leaves it to the renderer to figure it out. Though I'm curious how this might work with things like cancellation-on-click (for example, in vns when you click the text-writing-out animation finishes early), but I think it would just be on the renderer to report that the action is complete whether or not it's stopped early and it'd work fine? At least on the runner side, the renderer might have to handle cancelling animations/etc but the runner doesn't care.

>Different renderers would support different actions. For instance, the command line renderer would not support rendering sound, thus, the action would complete immediately as if it was performed. Or, maybe, as mentioned in the thread (I think), each renderer could provide a set of actions it supports and the runner would simply ignore certain nodes of the AST based on this knowledge.

Either way would work fine, but defining a list of valid renderer actions as part of the interface may come in handy in the future, and I lean towards pushing things into the runner rather than the renderer. I imagine something similar would come from the parser side, and this means the runner could compare the featureset of the parser between that and the runner. It doesn't really matter if the renderer supports more features than the parser, but it may be useful for a developer to see unusable renderer functions for the given parser.

I really like where this discussion is going, I'm starting to understand how this might lay itself out and I'm feeling good about it ^~^
